### DISCUSSION TITLE
How to import a project using a given Scala version

### DISCUSSION THREAD

**Discussion by adamw** [scalameta/metals#6784] [Q&A] [answered]
In Tapir I'm using [sbt-projectmatrix](https://github.com/sbt/sbt-projectmatrix) for cross-building, and many projects are enabled for multiple java versions/platforms. However, in the IDE I'd like to have a specific version+platform combination imported, so that I can work with a given source set etc.

Can I do this with metals? I've got:

```
val scala3 = "3.3.3"
val ideScalaVersion = scala3

  ideSkipProject := (scalaVersion.value != ideScalaVersion) ||
    thisProjectRef.value.project.contains("Native") ||
    thisProjectRef.value.project.contains("JS"),
  bspEnabled := !ideSkipProject.value,
```
 
But still the IDE seems to import the Scala 2 version of the code. Is there some SBT setting which I'm missing?

**Comment by tgodzik** [comment]
There is no way to do it currently and I actually wanted to work on that. I moved https://github.com/scalameta/metals/issues/6782 to the repo and will try to figure out a way for it there.

The manual way when using Bloop is to do remove the json file corresponding to the targets you don't want to use or there is similar workaround in the compiler https://github.com/scala/scala3/blob/65a53b5d98a84ffbc2b4e7a734639dc64c1a5b39/project/NoBloopExport.scala#L6
which you can use to disable importing specific targets.

When using sbt BSP bspEnabled should be enough.

**Comment by adamw** [comment]
So ... there is no way to do it, or `bspEnabled` should do the trick? :) I'm importing the project by clicking "import project" in VSCode, btw.

**Reply by tgodzik** [reply]
For sbt BSP bspEnabled should do the trick, but not for Bloop.

**Reply by adamw** [reply]
Hm sorry I don't understand ;) From the perspective of the user of the IDE, should this work or not? In SBT, when I do a build, everything is compiled in all versions/platforms as it should be.

Thie might be a stupid question, but the relation between Metals, BSP, VSCode and Bloop was always very confusing for me ;)

**Reply by tgodzik** [reply]
By default Metals will import build from sbt to Bloop and create the json files in `.bloop`. Each one corresponds to a compilable target. If we want to compile less, we need to force the import command to not generate some of them or remove those json files manually.

You can also switch to using sbt directly and it will respect the bspEnabled flag.

I can also look to making Bloop plugin respect that.

**Reply by adamw** [reply]
Ok thanks, that's a bit more clear now ... but what do you mean by "using sbt directly"? Is that for importing the project - so either somehow "use sbt directly" (invoke some goal?) as opposed to clicking "import project" - I thought this invokes sbt to generate the necessary files?

**Reply by tgodzik** [reply]
Yes, but sbt can also serve as a BSP server itself, though it works a bit different than bloop. This should be still valid: https://scalameta.org/metals/blog/2020/11/06/sbt-BSP-support/

**Reply by adamw** [reply]
Thanks, though ideally I'd just want to import my SBT project using Metals and to not have to worry about choosing Bloop/SBT etc. ;) I didn't change any defaults, so I guess I'm using Bloop? At least it appears sometimes in the status bar.

**Reply by tgodzik** [reply]
Normally Bloop should be good and I think it was overall better experience. It will be shown in the status bar as soon as it is connected.

**Comment by tgodzik** [comment, marked as answer]
Btw. this https://github.com/scalacenter/bloop/pull/2446 should make it work with just bspEnabled set

**Reply by adamw** [reply]
Ah great, thanks! :)
